File: Java/234. Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ArrayList<Integer> arr = new ArrayList<>();

        ListNode cur = head;
        while (cur != null){
            arr.add(cur.val);
            cur = cur.next;
        }
        int l = 0;
        int r = arr.size()-1;
        while (l<r){
            if (arr.get(l) != arr.get(r)){
                return false;
            }
            else{
                l++;
                r--;
            }
        }
        return true;
    }
}

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        //two pointers - fast and slow
        ListNode slow = head, fast = head;

        while (fast != null && fast.next!=null){
            fast = fast.next.next;
            slow = slow.next;
        } // slow will poiint the midddle and fast will point the end

        //now reverse the 2nd half
        ListNode prev = null;
        while (slow!=null){
            ListNode temp = slow.next;
            slow.next = prev;
            prev = slow;
            slow = temp;
        }
        // check for palindrome

        ListNode left = head, right = prev;
        while(right != null){
            if (left.val != right.val){
                return false;
            }
            left = left.next;
            right = right.next;
        }
        return true;
    }
}

File: Python/234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        arr = []
        curr= head
        while curr:
            arr.append(curr.val)
            curr=curr.next
        l = 0
        r = len(arr)-1
        while (l<r):
            if (arr[l]!=arr[r]):
                return False
            l+=1
            r-=1
        return True 

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        #two pointer - fast and slow
        fast = head
        slow = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        #fast will be at end, slow will be at middle

        #reverse 2dn half
        prev = None
        while slow:
            temp = slow.next
            slow.next=prev
            prev= slow
            slow = temp

        #check for palindrom
        l,r = head,prev

        while r:
            if (l.val != r.val):
                return False
            l = l.next
            r = r.next
        return True